### PR TITLE
Preliminary commits before session-based health checks

### PR DESCRIPTION
**Pull request by bwester**
This PR contains some preliminary commits that need to be in place before I introduce the main PR for session-based health checks. Those changes aren't quite ready for a PR, but I think the changes here are of high quality and not likely to change, so I'd like y'all to start looking these over.

I recommend reviewing one commit at a time, rather than looking at the entire "files changed" view for the PR.

### CODE REVIEW

**Comment by mpuncel** [comment]
lgtm, i think yours might conflict with my PR on the consul test fixture stuff but not a big deal

File: pkg/kp/kv.go [hunk 1]
```diff
@@ -52,6 +52,26 @@ type WatchResult struct {
 	Status  string
 	Output  string
 	Time    time.Time
+	Expires time.Time `json:"expires,omitempty"`
```
Comment (robertabbott) [inline pkg/kp/kv.go]: Should the other fields have json tags as well? Why just this one?

Comment (bwester) [inline pkg/kp/kv.go]: I added this field comment so I could specify the "omitempty" option, indicating that this field might not be present. The default JSON marshaller handles the other fields well enough; they don't need special handling.

But I do need a change here. As mentioned in another PR, the JSON field name should be capitalized, `Expires`, to match the other fields.  Interestingly, the YAML marshaller adjusts the case of its fields, and I was expecting that behavior from JSON too.

File: pkg/kp/kv.go [hunk 2]
```diff
@@ -52,6 +52,26 @@ type WatchResult struct {
...
+		r.Output == s.Output
+}
+
+// IsStale returns true when the result is stale according to the local clock.
+func (r WatchResult) IsStale() bool {
+	expires := r.Expires
+	if expires.IsZero() {
+		expires = r.Time.Add(TTL)
+	}
+	return time.Now().After(r.Expires)
```
Comment (bwester) [inline pkg/kp/kv.go]: Oops, should be `expires`.

**Comment by bwester** [comment]
Rebased to address review feedback

File: pkg/util/stream/publisher.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	case s.values <- val:
+		close(s.initialized)
+	case <-s.canceled:
+	}
+}
+
+// close the subscription's data stream if the subscription wasn't previously canceled
+func (s *StringSubscription) close() {
+	select {
+	default:
```
Comment (robertabbott) [inline pkg/util/stream/publisher.go]: was putting the default case first done for style/readability reasons?

Comment (bwester) [inline pkg/util/stream/publisher.go]: The main job of the method is to close its output socket, so I wrote that first. It's also makes the method similar to `waitForReady()` and `sendFirst()` in that checking for cancellation is always written last.

**Comment by petertseng** [comment]
I think I can get behind this PR, so 👍

minor nit for using `Fatal` rather than `Fatalln`, which I trust you'll resolve as you see fit

**Comment by bwester** [comment]
Amended to use `Fatalln()` per project convention instead of `Fatal()`.

**Pull request merged**